### PR TITLE
Revert "Move batch_norm to ATen/native, speed up (#12368)"

### PR DESCRIPTION
**Pull request by houseroad**
Revert #12368 since it's causing onnx related test cases failing.
https://github.com/pytorch/pytorch/pull/12368

@SsnL @t-vi

### CODE REVIEW

**Review by bddppq — APPROVED** [review]

**Comment by SsnL** [comment]
Please revert to unbreak master. Thank you!

**Comment by houseroad** [comment]
@t-vi here is the log of failing test: https://circleci.com/gh/pytorch/pytorch/129750?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by t-vi** [comment]
So I merge master into the my batch norm branch, built and installed onnx from third_party/onnx and ran
```
PYTHON_PATH=$(pwd)/build/lib.linux-x86_64-3.6/ python3 test/onnx/test_pytorch_onnx_caffe2.py TestCaffe2Backend.test_instance_norm
PYTHON_PATH=$(pwd)/build/lib.linux-x86_64-3.6/ python3 test/onnx/test_models.py TestModels.test_dcgan_netG TestModels.test_inception
```
and couldn't get the failure locally. :(

**Comment by houseroad** [comment]
@t-vi also forgot to mentioning, thanks a lot for the contribution, it's very useful. we just temporarily revert it to unbreak the CI test, I will help you push it back.

**Comment by t-vi** [comment]
I do appreciate your help with reproducing it and once I get to that, I hope we'll be able to fix it soon, too.

**Comment by houseroad** [comment]
No worries, I have uploaded the docker to my repo: https://hub.docker.com/r/houseroad/onnx-docker/tags/

```
docker pull houseroad/onnx-docker:repro
docker run -it houseroad/onnx-docker:repro /bin/bash
```

after launching docker, try
```
cd pytorch
python test/onnx/test_models.py TestModels.test_dcgan_netG
```

**Comment by t-vi** [comment]
Thank you for this! I can reproduce the error. Whyever it seemed a good idea at the time, not using sum inside parallel_for seems to fix the problem:
```
diff --git a/aten/src/ATen/native/Normalization.cpp b/aten/src/ATen/native/Normalization.cpp
index ca5f4ae..1774ec6 100644
--- a/aten/src/ATen/native/Normalization.cpp
+++ b/aten/src/ATen/native/Normalization.cpp
@@ -69,9 +69,12 @@ std::tuple<Tensor,Tensor,Tensor> batch_norm_cpu_template(const Tensor& input, co
 
 	if (train) {
 	  // compute mean per input
-	  accscalar_t sum = in.sum()._local_scalar().to<accscalar_t>(); // accumulation dtype ?
+	  accscalar_t sum = 0;
+	  CPU_tensor_apply1<scalar_t>(in, [&] (const scalar_t& i) {
+	      sum += i;
+	    });
 
-	  mean = (scalar_t) sum / n;
+	  mean = (scalar_t) (sum / n);
 	  save_mean_a[f] = mean;
 
 	  // compute variance per input
```
If it does for you, too, I'd have another go at the PR.

**Comment by houseroad** [comment]
@t-vi Awesome, it works for me on the local machine.

**Comment by soumith** [comment]
cc: @colesbury that seems like a bug with sum_parallel more than anything else?
